rust/src/mqtt/detect.rs

```diff
@@ -0,0 +1,481 @@
+            if let Some(p) = &cv.will_message {
+                if p.len() > 0 {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                    return 1;
+                }
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_connack_sessionpresent(
+    tx: &MQTTTransaction,
+    session_present: *mut bool,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNACK(ref ca) = msg.op {
+            *session_present = ca.session_present;
+            return 1;
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_publish_topic(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.topic;
+            if p.len() > 0 {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+                return 1;
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_publish_message(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.message;
+            if p.len() > 0 {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+                return 1;
+            }
+        }
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_has_topic(
+    tx: &MQTTTransaction,
+    result: *mut bool,
+    topic: *const std::os::raw::c_char,
+) -> u8 {
+    *result = false;
+    for msg in tx.msg.iter() {
+        match msg.op {
+            MQTTOperation::SUBSCRIBE(ref subv) => {
+                if let Ok(c_topic) = CStr::from_ptr(topic).to_str() {
+                    for topic in subv.topics.iter() {
+                        if topic.topic_name == c_topic {
+                            *result = true;
+                            return 1;
+                        }
+                    }
+                }
+            }
+            MQTTOperation::UNSUBSCRIBE(ref unsubv) => {
+                if let Ok(c_topic) = CStr::from_ptr(topic).to_str() {
+                    for topic in unsubv.topics.iter() {
+                        if topic == c_topic {
+                            *result = true;
+                            return 1;
+                        }
+                    }
+                }
+                return 0;
+            }
+            _ => return 0
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_reason_code(
```

Comment:
Do we really need `unsafe` for this function ?

Reply:
Well, it dereferences a raw pointer from C (`result`) to write the extracted value. This wouldn't work without `unsafe`, right?

```
error[E0133]: dereference of raw pointer is unsafe and requires unsafe function or block
   --> src/mqtt/detect.rs:437:21
    |
437 |                     *result = rcode;
    |                     ^^^^^^^^^^^^^^^ dereference of raw pointer
    |
    = note: raw pointers may be NULL, dangling or unaligned; they can violate aliasing rules and cause data races: all of these are undefined behavior
```

Reply:
Indeed, but you may have `unsafe` only for the dereference, and not the whole function.

Reply:
True. Will move the `unsafe` into the inner parts of the  code. Some of the functions there don't  need `unsafe` at all.